### DISCUSSION TITLE
How to develop the " setRC.tcl "  file from a captable for a foreign PDK ?

### DISCUSSION THREAD

**Discussion by faisaladilquadery** [The-OpenROAD-Project/OpenROAD#3943] [Q&A] [answered]
I have a foreign PDK which I want to invoke in openroad. But for that I need to develop the setRC.tcl file. How do I accurately input the resistance and capacitance for the setRC.tcl file . 

N.B : I have a captable for the pdk
I can also generate a spef file from Innovus for the PDK

**Comment by maliberty** [comment, marked as answer]
@vvbandeira @luarss do we have any documentation for correlateRC ?

Note that if you LEF res & cap has values you can use those without a setRC to get started.

**Comment by vvbandeira** [comment]
@faisaladilquadery 
We have a section on the `setRC.tcl` file configuration in our platform bring-up documentation.
https://openroad-flow-scripts.readthedocs.io/en/latest/contrib/PlatformBringUp.html#setrc-configuration

**Comment by faisaladilquadery** [comment]
Thankyou! @maliberty @vvbandeira . I have used the lef file in replacement of the setRC.tcl file . But I am getting **net delay** mismatch between the same path with the same cells for Innovus and OpenRoad. Any Idea why ?

**Reply by maliberty** [reply]
I would construct a simple two inverter test case to make it easy to compare the results.  Write a spef from both tools and see how they compare.

**Reply by faisaladilquadery** [reply]
I have already constructed a similar test case and discovering some discrepency in net delay values between the tools 
Can you guide me how I can generate the **spef** file in openroad and how can i generate the extraction rules file for my pdk?

**Reply by maliberty** [reply]
If you are using ORFS then a spef is automatically generated during the run.

For RCX see https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/rcx/doc/calibration.md
(@vvbandeira @luarss this should be linked from PlatformBringUp.html)

**Reply by faisaladilquadery** [reply]
Hi maliberty, I was working on a small test case with a input and output path, and I am noticing the the set_wire_rc 'RC' values are taken from the **last defined layer,** in this case, here only the metal 8 is being taken for the signal layer delay calculation, but not the other layers like metal 3 . How can i make the tool take signal rc values for multiple layers and calculate net delay for paths that have other metal layer besides layer 8?
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/140678965/002d4dba-0612-44e9-94c7-2d82028367da)

**Reply by maliberty** [reply]
set_wire_rc is used to set a default layer to use when doing placement based parasitic estimation.  There is no meaning to multiple layers here as there would be no way to divide up the estimate meaningfully.

**Reply by faisaladilquadery** [reply]
I see,  does the ORFS still take the estimate values during net delay values in the route stage ?
and which formula or method does the tool use to calculate the net delay from the RC values provided by me ?
Thanks in advance :)

**Reply by maliberty** [reply]
Once you have global routes we use the per layer values from set_layer_rc.  At the end of the flow we use rcx and its tech file for the most accurate results.

**Comment by faisaladilquadery** [comment]
Thankyou @maliberty I have used the the RCX flow you  mentioned to generate a rc tech file for my pdk and used the tech file to generate a spef in ORFS to obtain a more accurate RC.

**Comment by b224hisl** [comment]
> set_wire_rc is used to set a default layer to use when doing placement based parasitic estimation. There is no meaning to multiple layers here as there would be no way to divide up the estimate meaningfully.

So I'm confused why it needs to set jsut one layer, do you mean only placement based parasitic estimation is related to `set_wire_rc` ? If there' s no way to divide up the estimate meaningfully, the layer can be set randomly?

**Reply by maliberty** [reply]
Yes placement based estimates use a single layer from set_wire_rc.  I wouldn't it should be set randomly but rather to a representative layer of the common case.
